Six frequency support

ourairquality opened the pull request:
Submitting this set again in case it helps now. The naming of the options may well want to be changed, and there is no attempt to add support for the windows app, but perhaps this is a guide as to what might need updating.

ourairquality commented:
Fwiw, merged in the updates to the winapp but with some modifications. Attempted to extend this to 6 frequencies (gui layout is likely wrong) and also suggesting a different naming of some of the options see below for discussion.

Noted that the qt app did not appear to be built for 4 frequencies, like the winapp now, and RTKLib.pri needed updating, and the SNR mask support appeared to only support 3 frequencies at present.

**Option naming**

There are two sets of options with names that are for respective frequencies indices but are numbered in the sequence 1,2,5 and now 1,2,5,6. These are `pos1-snrmask_L{1,2,5,6} `and `stats-eratio{1,2,5,6}`. The GUI apps also offer the number of frequencies options as L1+L2+L5+L6 for 4 frequencies. While this naming worked for GPS, QZSS and SBAS, it does not fit for the other systems and does not extend well to 6 frequencies.

This PR proposes to just number and name these sequentially as a frequency index as that is how the code currently uses these. It proposes to name these `pos1-snrmask{1,2,3,4,5,6}` and `stats-eratio{1,2,3,4,5,6}`. This is an incompatible renaming for the SNR mask, and also for `stats-eratio5` which would become `stats-eratio3`.

The stats-eratio is used in the function `varerr()` which also has heuristics for different systems and signals. There may well be a need to have a different errors for each system signal, but that is left for future work on varerr(), perhaps as coded heuristics or perhaps as more config options.

rtklibexplorer commented:
At least for now, I'm leaning towards supporting six frequencies for the conversion and plotting apps but limiting the solution apps to four frequencies.  I think it will be fairly easy to add a frequency mapping layer through a simple text file (similar to the way the station position file is handled) to allow the advanced user to choose which four frequencies are used for each constellation while avoiding additional complexity for the average user.  This will also minimize rework to the GUIs which can be tedious and time-consuming.  Agreed that there is still some cleanup needed regarding inconsistent use of L1/L2/L3/L4 and L1/L2/L5/L6.  I'm thinking we should settle on L1/L2/L3/L4 but with more use of the freqdlg button (the ?) to make it easier for the user to display the current freq mapping from any relevant menu.

ourairquality committed:
winapp rtknavi: remove dup labels, update tab order, fix some options

Remove some duplicate labels or unused labels.

Improve the tab orderings.

Reorganise the GUI layout descriptions to be closer to the order of
presentation, closer to the tab ordering. The development tools do not
maintain an order, but it is tedious to try to use the graphical
development tools to obtain consistent layout.

Improve the layout alignments.

Add file loading for the BLQ file.

Fix the loading and saving of the BLQ and proxy address config options.

Fix viewing of the DCB file. The method was missing and it was using
the station position file.

Add rtklib.h to a few header files.

Correct solution sstat and trace options. The state for these was
duplicated in other variables and not always being set in the solopt_t
object. Remove this duplication to resolve this.

rtkconv: default to RINEX 3.04, and enable all signal types

rtkpost: increase the number of input files from 6 to 10, matching the
qt app.

Update the frequency dialog to include the 6th index and BDS B2ab.

srvtblbrows: Expand the cast list width to better view long
URLs. Strip trailing carriage returns from URLs.

Remove some unused variables, to quieten compiler warnings.

ourairquality commented:
Fwiw updated the winapp support. Have setup a windows compiler environment to actually test the proposals now.

Not sure about manual frequency mapping approach. For a particular setup that might work, with particular rover and base hardware, but it might be a barrier to general use. I have been working towards avoiding such an approach and supporting all available signals in part of that.

The RTK solution using the continuous multi-frequency method for ambiguity resolution appears to be robust to using all available signals (one common signal between the rover and base for each frequency index). When RTKLIB is configured for many solution frequency indices it usually just works without having to select the particular signals to use. If there is no common signal at a frequency index then it is just not used in the solution. With this approach it is possible to pick RTCM streams and have a good chance of getting a good solution using all available signals, and that seems less of a burden for the user. E.g. tested this PR with a selection of local fixed sites with modest baselines, quality receivers sending out a lot of signals, some IGS reference sites, and got a fix.

But I agree that there should be a way to select the signals, but this might not need to affect the signal to frequency index mapping, rather it might be a selection from a hard coded signal to frequency index mapping. For example to use BDS B1 B2 B3 in a solution the user could select frequency indices F1 F2 and F4 (currently aka L1 L2 L6). Not saying this is the best approach. There is already some code doing such a mapping, see `seliflc()`, and this could be a path to allow signal selection too, to allow this mapping to be changed by user configuration options.

This PR helps show what it looks like to name the frequency indices as F1, F2, F3, F4, F5, F6 rather than L1, L2, L5, L6, L7, L8; and also to enable support for 6 frequency indices.

PS there are more changes that could help solutions using many frequency indices, proposals dropped for now, such as having a common signal index function with more flexibly signal mapping, and some experimental changes to the RTK solution to not confuse signals across systems at the same frequency index, improvements to the antex support to better support the range of antenna frequencies, etc.  But this does not appear to be a reason to hold back plumbing the support for more frequency indices as the user can trivially limit the solution to just 2,3,or 4 indices if necessary.

ourairquality commented:
Updated the rtkplot sat visibility plot to include the extra observation frequency indices, and have it limit the frequency indices and codes selection list to those for the selected satellites. So it is now possible to select a particular satellite and see just the frequency indices used and the signal codes available. Even though it supports many frequency indices, the selection only shows those available in the data and so e.g. just two for a dual frequency receiver.

Extended the 'All' signals output to count up to 6 observations, up from 5. This was named #FRQ in the winapp but it is counting the number of observations which can be more than the number of frequency indices when there are multiple observations per frequency index, so changed this to #OBS as in the qt app.

ourairquality committed:
options: expand to handle 6 frequency indices

Give an option for pos1-frequency to be up to 6 frequency indices,
adding values 5, and 6.

Expand the SNR masks to 6 frequency indices, up from 4. Rename the SNR
mask frequency options from pos1-snrmask_L{1,2,5} to and
pos1-snrmask{1,2,3,4,5,6}, so now referring to the frequency index
rather than attempting to refer to the signal.

Expand the eratio parameters to 6 frequency indices, up from
4. Reworking the option naming to refer to the signal index rather
than attempting to refer to the signal, so now
stats-eratio{1,2,3,4,5,6} for the respective frequency index.

Change pos1-frequency to be just an integer 1 to 6, and improve the
guard.

ourairquality commented:
Closing this one. Have taken into account the feedback above, to allow runtime configuration of the frequency index mapping, and have a suggestion that uses a config option to do so, but it is a large change and it supersedes the suggestions here so shall open a new topic for that.
